### PR TITLE
Fixed #28250 -- Ignored soft applied migrations in consistency check.

### PR DESCRIPTION
**Pull request by knbk**
Ignored initial migrations that have been soft applied and may be faked
with the --fake-initial flag in the migration history consistency
check. Does not ignore the initial migration if a later migration in
the same app has been recorded as applied.

Thanks to Raphaël Hertzog for the initial patch.

Ticket: https://code.djangoproject.com/ticket/28250

### CODE REVIEW

**Comment by knbk** [comment]
Buildbot, test this please

**Comment by knbk** [comment]
buildbot, test this please

**Comment by knbk** [comment]
buildbot, test this please

**Comment by brianmay** [comment]
Unfortunately, while this solves the initial error, we get another error later on.

```
Traceback (most recent call last):
            File "/usr/bin/lava-server", line 78, in <module>
              main()
            File "/usr/bin/lava-server", line 74, in main
              execute_from_command_line(django_options)
            File
          "/usr/lib/python2.7/dist-packages/django/core/management/__init__.py",
          line 367, in execute_from_command_line
              utility.execute()
            File
          "/usr/lib/python2.7/dist-packages/django/core/management/__init__.py",
          line 359, in execute
              self.fetch_command(subcommand).run_from_argv(self.argv)
            File
          "/usr/lib/python2.7/dist-packages/django/core/management/base.py", line
          294, in run_from_argv
              self.execute(*args, **cmd_options)
            File
          "/usr/lib/python2.7/dist-packages/django/core/management/base.py", line
          345, in execute
              output = self.handle(*args, **options)
            File
          "/usr/lib/python2.7/dist-packages/django/core/management/commands/migrate.py",
          line 164, in handle
              pre_migrate_apps = pre_migrate_state.apps
            File "/usr/lib/python2.7/dist-packages/django/utils/functional.py",
          line 35, in __get__
              res = instance.__dict__[self.name] = self.func(instance)
            File "/usr/lib/python2.7/dist-packages/django/db/migrations/state.py",
          line 176, in apps
              return StateApps(self.real_apps, self.models)
            File "/usr/lib/python2.7/dist-packages/django/db/migrations/state.py",
          line 249, in __init__
              raise ValueError("\n".join(error.msg for error in errors))
          ValueError: The field lava_scheduler_app.TestJob.submit_token was
          declared with a lazy reference to 'linaro_django_xmlrpc.authtoken', but
          app 'linaro_django_xmlrpc' isn't installed.
```

I am not entirely convinced that this change is causing this problem, however we have ruled out all the obvious suspects.

The ``linaro_django_xmlrpc`` is definitely listed in ``INSTALLED_APPS`` and the problem goes away if we manually insert the fake transaction into the database before running the transactions.

Full log: http://paste.debian.net/958558/

As tested by Senthil Kumaran S.

**Comment by timgraham** [comment]
I haven't studied this in much detail but I didn't spot an explanation about moving `detect_soft_applied()` to `MigrationLoader`. Could you give the reasoning?

**Comment by knbk** [comment]
`MigrationLoader.check_consistent_history()` needs to call `detect_soft_applied`, but the loader has no access to the executor. Moving the method solves this.

File: django/db/migrations/executor.py [hunk 1]
```diff
@@ -77,6 +76,24 @@ def _create_project_state(self, with_applied_migrations=False):
             for migration, _ in full_plan:
                 if migration in applied_migrations:
                     migration.mutate_state(state, preserve=False)
+                elif self.loader.is_initial_migration(migration):
```
Comment (carltongibson) [inline django/db/migrations/executor.py]: This block is hard. I think a comment would make it clearer. 

```
# Check if we need to fake apply an initial migration with applied children. 
```

File: django/db/migrations/executor.py [hunk 2]
```diff
@@ -77,6 +76,24 @@ def _create_project_state(self, with_applied_migrations=False):
...
+                    fake_initial = False
+                    queue = deque([migration])
+                    while queue and not fake_initial:
+                        m = queue.popleft()
+                        for child in self.loader.graph.node_map[m.app_label, m.name].children:
+                            child_migration = self.loader.graph.nodes[child]
+                            if child_migration in applied_migrations:
+                                fake_initial = True
+                                break
+                            if self.loader.is_initial_migration(child_migration):
```
Comment (carltongibson) [inline django/db/migrations/executor.py]: I wasn't 100% sure what was the purpose of this. (I didn't quite follow the discussion from the ticket of the cross-app dependency.) 

Two points: 

* Again a comment here would help. `# Re-apply across apps...` (or something)
* There's no test covering this case: If I comment out these lines nothing fails.

**Comment by LuRsT** [comment]
@knbk Need help fixing the comments from the code review?

**Comment by brianmay** [comment]
I am wondering if this change is still desirable?

**Comment by meshy** [comment]
I've just hit this issue. This patch is still useful for me.

**Comment by felixxm** [comment]
Closing due to inactivity.

**Pull request closed without merging**